operator/iamkeycontroller/create.go

```diff
@@ -0,0 +1,146 @@
+func (p *IAMKeyPipeline) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	log := controllerruntime.LoggerFrom(ctx)
+	log.Info("Creating resource")
+
+	iam := fromManaged(mg)
+	if iam.Status.AtProvider.KeyID != "" {
+		// IAMKey already exists
+		return managed.ExternalCreation{}, nil
+	}
+
+	pipe := pipeline.NewPipeline().WithBeforeHooks(steps.DebugLogger(ctx)).
+		WithSteps(
+			pipeline.NewStepFromFunc("create IAM key", p.createIAMKeyFn(iam)),
+			pipeline.If(hasSecretRef(iam),
+				pipeline.NewStepFromFunc("create credentials secret", p.createCredentialsSecretFn(iam)),
+			),
+			pipeline.NewStepFromFunc("emit event", p.emitCreationEventFn(iam)),
+		)
+	result := pipe.RunWithContext(ctx)
+	if result.IsFailed() {
+		return managed.ExternalCreation{}, errors.Wrap(result.Err(), "cannot create IAM Key")
+	}
+
+	return managed.ExternalCreation{ConnectionDetails: toConnectionDetails(p.exoscaleIAMKey)}, nil
+}
+
+// createIAMKeyFn creates a new IAMKey in the project associated with the API Key and Secret.
+func (p *IAMKeyPipeline) createIAMKeyFn(iamKey *exoscalev1.IAMKey) func(ctx context.Context) error {
+	return func(ctx context.Context) error {
+		exoscaleClient := p.exoscaleClient
+		log := controllerruntime.LoggerFrom(ctx)
+
+		var keyResources []v2.IAMAccessKeyResource
+		for _, bucket := range iamKey.Spec.ForProvider.SOS.Buckets {
+			keyResource := v2.IAMAccessKeyResource{
+				Domain:       SOSResourceDomain,
+				ResourceName: bucket,
+				ResourceType: BucketResourceType,
+			}
+			keyResources = append(keyResources, keyResource)
+		}
+
+		exoscaleIAM, err := exoscaleClient.CreateIAMAccessKey(ctx, iamKey.GetZone(), iamKey.GetKeyName(), v2.CreateIAMAccessKeyWithResources(keyResources))
+		if err != nil {
+			return err
+		}
+
+		if err != nil {
+			return err
+		}
+		// Limitation by crossplane: The interface managed.ExternalClient doesn't allow updating the resource during creation except annotations.
+		// But we need to somehow store the key ID returned by the creation operation, because exoscale API allows multiple IAM Keys with the same display name.
+		// So we store it in an annotation since that is the only allowed place to update our resource.
+		// However, once we observe the spec again, we will copy the key ID from the annotation to the status field,
+		//  and that will become the authoritative source of truth for future reconciliations.
+		metav1.SetMetaDataAnnotation(&iamKey.ObjectMeta, KeyIDAnnotationKey, *exoscaleIAM.Key)
+		metav1.SetMetaDataAnnotation(&iamKey.ObjectMeta, BucketsAnnotationKey, getBuckets(*exoscaleIAM.Resources))
+
+		log.V(1).Info("Created IAM Key in exoscale", "keyID", *exoscaleIAM.Key, "displayName", *exoscaleIAM.Name, "tags", exoscaleIAM.Tags)
+		p.exoscaleIAMKey = exoscaleIAM
+		return nil
+	}
+}
+
+func getBuckets(iamResources []v2.IAMAccessKeyResource) string {
+	var buckets strings.Builder
+	for _, iamResource := range iamResources {
+		buckets.WriteString(",")
```

Comment:
We can add an `if` but take in consideration that the webhook does not allow to create iam keys with no buckets thus no resources.

Reply:
At a later stage that behavior will change, when we support other services. Wouldn't it be possible then that an IAM key contains other resources as well?

Reply:
It will be possible to add other services but it will not affect the logic for the `sos service`. The only case this will not work is if we allow IAM keys to have access to all buckets thus leaving the allowed list an empty list. From my perspective this is not very desirable as it will create confusion from our client's perspective (why empty list should mean access to all buckets). Going forward with this idea, we actually create and IAM Key with an empty list and the provider would populate in real time all the buckets from into that list but we may have tens if not hundres of buckets in a project which again is not a perfect solution. In any case I updated the code and added the `if` as it adds safety and doesn't have any regression.